### ISSUE TITLE
Generalization of Selective Model Extension for Equations?

### ISSUE THREAD

**Issue by mestinso** [modelica/ModelicaSpecification#3363]
The new Modelica 3.6 "Selective Model Extension" feature seems quite useful when a base class is built graphically with components and connect statements. However, in situations where a base class is built primarily with equations. the usefulness is limited.

Is there any thought around either generalizing this feature to enable the removal of equations and/or adding an analogous feature for removal of equations?

Below is a trivial example use case and potential syntax:
```
model Example10
  Real a;
equation
  a = 10;
end Example10;

model Example5
  extends Example10(break equation a);
equation
  a = 5;
end Example5;
```

Since it's awkward/difficult to isolate specific equations, my thinking is that it might be most natural that **all** equations with the variable mentioned in the *equation break statement* should be removed.

**Comment by HansOlsson** [comment]
That haven't been given much thought, and I see some related issues:

- Equations can be nested in various forms - removing an entire 20-line for-loop due to one use of a variable in one line seems excessive, but cutting it into parts also seems problematic.
- You will anyway have to look at the base-class to see which variables occur in which specific equation, and it's not clear how to do that efficiently. Is `x` used directly or is there another equation computing `y=2*x` and using `y`? In one sense this is about visualizing the result - with graphics that is more straightforward.

Thus I wouldn't see this as a priority at the moment.

**Comment by christoff-buerger** [comment]
It certainly would be possible. The idea is to settle with the current design for now and see how well it is received, how it is used, hence to get some body of knowledge; having such we can extend as we see fit.

There are many ways to do it. An alternative to your proposal would be to deselect the whole variable and all usages of it. Hence, it could be more a "deselect declaration" than "deselect equation".

**Comment by mestinso** [comment]
> That haven't been given much thought, and I see some related issues:
> 
> * Equations can be nested in various forms - removing an entire 20-line for-loop due to one use of a variable in one line seems excessive, but cutting it into parts also seems problematic.

I agree with the concern, and it seems like the more conservative removal would be best. 

> * You will anyway have to look at the base-class to see which variables occur in which specific equation, and it's not clear how to do that efficiently. Is `x` used directly or is there another equation computing `y=2*x` and using `y`? In one sense this is about visualizing the result - with graphics that is more straightforward.

I agree with how it would be used (looking at the base class and intelligently removing equations to fit the situation+goal).

> There are many ways to do it. An alternative to your proposal would be to deselect the whole variable and all usages of it. Hence, it could be more a "deselect declaration" than "deselect equation".

For simple variables, yes I don't see much advantage one way or another. However, when the variables belong to a model/connector then I see a clear advantage to just removing the equation. For example, maybe the base class has a fluid port `fluidPort_a` and I want to remove any equations that have `fluidPort_a.m_flow` in them. And for the situations where I want to remove both declaration and equations, I can still do that (by utilizing the existing `break` functionality). Furthermore, standing back and looking at this feature combined with the existing features, it seems fairly comprehensive/complete/useful:
1) `extend Class(break [variable/component]...` to remove a declaration
2) `extend Class(break connect(...` to remove a connect equation
3) `extend Class(break equation [variable]...` to remove equations involving `variable`

Lastly, I would comment that I'm not just proposing this feature for "fun". I have various real-world examples where I'm using commercial libraries that are heavily equation based where this feature would be very handy. Currently, in those situations, I am having to duplicate the base class (typically 100+ lines of code), remove a few equations, and then add a few equations (typically <5 lines modified). With this feature, I would be able to extend (with the break statement), add a few equations, and be done with it. This would be a huge maintainability win!

**Comment by christoff-buerger** [comment]
> Lastly, I would comment that I'm not just proposing this feature for "fun". I have various real-world examples where I'm using commercial libraries that are heavily equation based where this feature would be very handy. Currently, in those situations, I am having to duplicate the base class (typically 100+ lines of code), remove a few equations, and then add a few equations (typically <5 lines modified). With this feature, I would be able to extend (with the break statement), add a few equations, and be done with it. This would be a huge maintainability win!

I assumed you have a good reason to aks for such feature. This sounds very good. We will look into it.

Considering the pace of MCP changes, please don't have high hopes this will be included in the next half year or anything like that :)
